core/trino-main/src/main/java/io/trino/sql/planner/iterative/Plans.java

```diff
@@ -46,6 +46,10 @@ protected PlanNode visitPlan(PlanNode node, Void context)
                     .map(child -> child.accept(this, context))
                     .collect(Collectors.toList());
 
+            if (node.getSources().equals(children)) {
+                return node;
+            }
```

Comment:
How is this related?

Reply:
It is needed for `resolvedPlan.equals(actual.getRoot()` to work IIUC.

Maybe instead doing this equality check you can explicitly check in `assertPlan` if `actual` contains `GroupReference` nodes or not?

Reply:
thought about such approach but decided it's cleaner this way

Reply:
I still don't understand why it's necessary. If `node.getSources()` is equal to `children`, then `node` should be equal to `node.replaceChildren(childen)` unless there's a difference in the node itself.

Reply:
From what I see we are not overriding the `equals` for `PlanNode` subclasses. So for `node` to be equal to `node.replaceChildren(childen)`, the method must return same object.

Reply:
In that case, this is very brittle. We're relying on identity comparison where there's no guarantee in the planner that identity is going to be preserved. If anything, it's an optimization, but we shouldn't rely on identity equality. If we need to compare plans in that manner, we need to ensure equals/hashcode for plan nodes is properly implemented (which is tricky to do efficiently because of the recursive nature of plan nodes)